staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,199 @@
+		klog.Error("invalid lease object provided to updatePeerServedResources()")
+		return
+	}
+
+	newLease, ok := newObj.(*v1.Lease)
+	if !ok {
+		klog.Error("invalid lease object provided to updatePeerServedResources()")
+		return
+	}
+
+	switch {
+	case newLease == nil && oldLease == nil:
+		return
+	case newLease != nil && oldLease != nil:
+		// Delete old discovery info if holderIdentity changed, implying
+		// the sever was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.peerAggDiscoveryCacheLock.Lock()
+			delete(h.peerAggDiscoveryResponseCache, oldLease.Name)
+			h.peerAggDiscoveryCacheLock.Unlock()
+		}
+
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease == nil && newLease != nil:
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease != nil && newLease == nil:
+		h.peerAggDiscoveryCacheLock.Lock()
+		delete(h.peerAggDiscoveryResponseCache, oldLease.Name)
+		h.peerAggDiscoveryCacheLock.Unlock()
+	}
+}
+
+func (h *peerProxyHandler) addPeerDiscoveryInfoToCache(serverIdentityLease *v1.Lease) error {
+	// Don't record our own discovery info.
+	if serverIdentityLease.Name == h.serverId {
+		return nil
+	}
+
+	h.peerAggDiscoveryCacheLock.RLock()
+	discoveryDoc, ok := h.peerAggDiscoveryResponseCache[serverIdentityLease.Name]
+	h.peerAggDiscoveryCacheLock.RUnlock()
+	if !ok {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	if discoveryDoc.holderIdentity != *serverIdentityLease.Spec.HolderIdentity {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	return nil
+}
+
+func (h *peerProxyHandler) fetchNewDiscoveryFor(serverID string, holderIdentity *string) error {
+	discoveryDoc := &peerAggDiscoveryInfo{
+		holderIdentity:  *holderIdentity,
+		servedResources: make(map[schema.GroupVersion][]string),
+	}
+
+	hostport, err := h.hostportInfo(serverID)
+	if err != nil {
+		return fmt.Errorf("failed to get host port info from identity lease for server %s: %w", serverID, err)
+	}
+
+	klog.V(4).Infof("Proxying an agg-discovery call from %s to %s", h.serverId, serverID)
+	discoveryPaths := []string{"/api", "/apis"}
+	for _, path := range discoveryPaths {
+		discoveryResponse, err := h.aggregateDiscovery(path, hostport)
+		if err != nil {
+			klog.ErrorS(err, "error querying discovery endpoint", "path", path, "serverID", serverID)
+			continue
+		}
+
+		if discoveryResponse == nil {
+			klog.V(4).InfoS("discovery response is nil", "path", path, "serverID", serverID)
+			continue
+		}
+
+		for _, groupDiscovery := range discoveryResponse.Items {
+			groupName := groupDiscovery.Name
+			if groupName == "" {
+				groupName = "core"
+			}
```

Comment:
Curious why we're switching to "core" for group here.  Is there a problem with using the convention of "" to indicate the core group?  Would a comment be helpful to future readers?

Reply:
Hmm.. I was doing the [same thing](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go#L142-L144) in the initial version of MVP for the incoming resource request so kept it the same way here too..

I'll check if this can be omitted